module/state_synchronization/indexer/persister.go

```diff
@@ -0,0 +1,221 @@
+	"go.uber.org/multierr"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+	"github.com/onflow/flow-go/storage/store/inmemory/unsynchronized"
+	"github.com/onflow/flow-go/utils/logging"
+)
+
+// Persister handles transferring data from in-memory storages to permanent storages.
+type Persister struct {
+	log zerolog.Logger
+
+	inMemoryRegisters      *unsynchronized.Registers
+	inMemoryEvents         *unsynchronized.Events
+	inMemoryCollections    *unsynchronized.Collections
+	inMemoryTransactions   *unsynchronized.Transactions
+	inMemoryResults        *unsynchronized.LightTransactionResults
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages
+
+	registers      storage.RegisterIndex
+	events         storage.Events
+	collections    storage.Collections
+	transactions   storage.Transactions
+	results        storage.LightTransactionResults
+	txResultErrMsg storage.TransactionResultErrorMessages
+	protocolDB     storage.DB
+
+	executionResult *flow.ExecutionResult
+	header          *flow.Header
+}
+
+// NewPersister creates a new persister.
+func NewPersister(
+	log zerolog.Logger,
+	inMemoryRegisters *unsynchronized.Registers,
+	inMemoryEvents *unsynchronized.Events,
+	inMemoryCollections *unsynchronized.Collections,
+	inMemoryTransactions *unsynchronized.Transactions,
+	inMemoryResults *unsynchronized.LightTransactionResults,
+	inMemoryTxResultErrMsg *unsynchronized.TransactionResultErrorMessages,
+	registers storage.RegisterIndex,
+	events storage.Events,
+	collections storage.Collections,
+	transactions storage.Transactions,
+	results storage.LightTransactionResults,
+	txResultErrMsg storage.TransactionResultErrorMessages,
+	protocolDB storage.DB,
+	executionResult *flow.ExecutionResult,
+	header *flow.Header,
+) *Persister {
+	log = log.With().
+		Str("component", "persister").
+		Hex("block_id", logging.ID(executionResult.BlockID)).
+		Uint64("height", header.Height).
+		Logger()
+
+	persister := &Persister{
+		log:                    log,
+		inMemoryRegisters:      inMemoryRegisters,
+		inMemoryEvents:         inMemoryEvents,
+		inMemoryCollections:    inMemoryCollections,
+		inMemoryTransactions:   inMemoryTransactions,
+		inMemoryResults:        inMemoryResults,
+		inMemoryTxResultErrMsg: inMemoryTxResultErrMsg,
+		registers:              registers,
+		events:                 events,
+		collections:            collections,
+		transactions:           transactions,
+		results:                results,
+		txResultErrMsg:         txResultErrMsg,
+		executionResult:        executionResult,
+		header:                 header,
+		protocolDB:             protocolDB,
+	}
+
+	persister.log.Info().
+		Uint64("height", header.Height).
+		Msg("persister initialized")
+
+	return persister
+}
+
+// Persist save data from in-memory storages to the provided persisted storages and commit updates to the database.
+// No errors are expected during normal operations
+func (p *Persister) Persist() (err error) {
```

Comment:
It's possible this Persist might be called more than once, maybe we could have a way to prevent that.